h/storage.py

```diff
@@ -180,18 +202,20 @@ def update_annotation(request, id_, data):
             "group: " + _("Invalid group specified for annotation")
         )
 
-    if document:
+    if document or uri_changed:
         annotation.document = update_document_metadata(
             request.db,
             annotation.target_uri,
-            document["document_meta_dicts"],
-            document["document_uri_dicts"],
+            document.get("document_meta_dicts", {}),
+            document.get("document_uri_dicts", {}),
             updated=annotation.updated,
         )
 
     request.find_service(  # pylint: disable=protected-access
         name="search_index"
-    )._queue.add_by_id(annotation.id, tag="storage.update_annotation", schedule_in=60)
+    )._queue.add_by_id(
+        annotation.id, tag=reindex_tag, schedule_in=0 if reindex else 60, force=reindex
```

Comment:
```suggestion
        annotation.id, tag=reindex_tag, force=not update_timestamp
```

I don't think you need the `schedule_in` change here. As explained above `schedule_in=0` won't cause the reindexing to happen immediately. It will still take up to 30s or longer if there are more than 400 other jobs ahead of it on the queue. The job queue never does anything immediately. You don't actually need the `60s` delay in your case but for code simplicity I think you may as well leave it in.

For `force` I think you can do `force=not update_timestamp` and so avoid the need for the `reindex` argument.